extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts

```diff
@@ -0,0 +1,402 @@
+	public databaseNameTextbox: azdata.InputBoxComponent | undefined;
+	public databaseVariableTextbox: azdata.InputBoxComponent | undefined;
+	public serverNameTextbox: azdata.InputBoxComponent | undefined;
+	public serverVariableTextbox: azdata.InputBoxComponent | undefined;
+
+	public currentReferenceType: ReferenceType | undefined;
+	private referenceLocationMap: Map<string, DatabaseReferenceLocation>;
+
+	private toDispose: vscode.Disposable[] = [];
+	private initDialogComplete: Deferred<void> | undefined;
+	private initDialogPromise: Promise<void> = new Promise<void>((resolve, reject) => this.initDialogComplete = { resolve, reject });
+
+	public addReference: ((proj: Project, settings: ISystemDatabaseReferenceSettings | IDacpacReferenceSettings) => any) | undefined;
+
+	constructor(private project: Project) {
+		this.dialog = azdata.window.createModelViewDialog(constants.addDatabaseReferenceDialogName);
+		this.addDatabaseReferenceTab = azdata.window.createTab(constants.addDatabaseReferenceDialogName);
+
+		this.referenceLocationMap = new Map([
+			[constants.sameDatabase, DatabaseReferenceLocation.sameDatabase],
+			[constants.differentDbSameServer, DatabaseReferenceLocation.differentDatabaseSameServer],
+			[constants.differentDbDifferentServer, DatabaseReferenceLocation.differentDatabaseDifferentServer]
+		]);
+	}
+
+	public async openDialog(): Promise<void> {
+		this.initializeDialog();
+		this.dialog.okButton.label = constants.addDatabaseReferenceOkButtonText;
+		this.dialog.okButton.enabled = false;
+		this.toDispose.push(this.dialog.okButton.onClick(async () => await this.addReferenceClick()));
+
+		this.dialog.cancelButton.label = constants.cancelButtonText;
+
+		azdata.window.openDialog(this.dialog);
+		await this.initDialogPromise;
+	}
+
+	private dispose(): void {
+		this.toDispose.forEach(disposable => disposable.dispose());
+	}
+
+	private initializeDialog(): void {
+		this.initializeTab();
+		this.dialog.content = [this.addDatabaseReferenceTab];
+	}
+
+	private initializeTab(): void {
+		this.addDatabaseReferenceTab.registerContent(async view => {
+			this.view = view;
+			const radioButtonGroup = this.createRadioButtons();
+			this.systemDatabaseFormComponent = this.createSystemDatabaseDropdown();
+			this.dacpacFormComponent = this.createDacpacTextbox();
+			const locationDropdown = this.createLocationDropdown();
+			const variableSection = this.createVariableSection();
+
+			this.formBuilder = <azdata.FormBuilder>view.modelBuilder.formContainer()
+				.withFormItems([
+					{
+						title: '',
+						components: [
+							radioButtonGroup,
+							this.systemDatabaseFormComponent,
+							locationDropdown,
+							variableSection
+						]
+					}
+				], {
+					horizontal: false
+				})
+				.withLayout({
+					width: '100%'
+				});
+
+			let formModel = this.formBuilder.component();
+			await view.initializeModel(formModel);
+
+			this.initDialogComplete?.resolve();
```

Comment:
initDialogComplete- this variable will be there always- why a ?

Reply:
initDialogComplete is Deferred<void> | undefined because it isn't initialized in the constructor